### ISSUE TITLE
Page-End-Hooks are executed more than once

### ISSUE THREAD

**Issue by BinkaDroid** [friendica/friendica#7001]
For example: The impressum-addon lets you setup a footnote (I use this, to give a friend a "help-button").

### Expected behavior
one footnote per page

### Actual behavior
Up to 3 footnotes per page (resulting in invalid html, as there are multipe instances of the same ID)

![hookbug](https://user-images.githubusercontent.com/39274494/56053120-86fe6b00-5d53-11e9-88da-296d6902dfdf.png)

### Steps to reproduce the problem
- enable impressum-addon
- setup a footnote
- create new message
- open file-browser & click on "files"

Using "frio", you can see 3 footnotes; in "vier" they are not visible, but exist in the sourcecode of the page

This is, how the addon registers its hook:
`Hook::register('page_end', 'addon/impressum/impressum.php', 'impressum_footer');`

### Expected behavior
'page_end' indicates, that this hook would only be executed once per page

### Actual behavior
sometimes (eg in filebrowser) it is executed more than once

I checked other addons and they have the same problem.

Example (cookie notice):
![hookbug2](https://user-images.githubusercontent.com/39274494/56053139-9382c380-5d53-11e9-9282-d751a5746b7c.png)
Just one notice is visible in browser, but source contains 3 instances of that ID ...

With addon/buglink it's the same ...

So there seems to be a general problem with how often page-end-hooks are executed under certain conditions, eg. when using filebrowser.

As I am a sub-sub-sub-junior, this is way over my head to fix ;-)

For my frindica-installation I "fixed" it the dirty way:

```
function impressum_footer($a, &$b) {
    $text = ProxyUtils::proxifyHtml(BBCode::convert(Config::get('impressum','footer_text')));

    if (! $text == '') {
        $a->page['htmlhead'] .= '<link rel="stylesheet" type="text/css" href="'.$a->getBaseURL().'/addon/impressum/impressum.css" media="all" />';
		/* dirty hack to prevent multiple instances of footer-text in fbrowser of editor */
		if (!strpos($b, 'fbrowser')) {
			/*hack end */
        $b .= '<div class="clear"></div>';
        $b .= '<div id="impressum_footer">'.$text.'</div>';
		}
    }
}
```

... but that is clearly not a fix for the more general problem.

(I hope my english isn't that bad ;-) )

**Comment by MrPetovan** [comment]
Your english is perfectly fine. The file browser is a standalone page that's asynchronously loaded in a modal window, so it would make sense if there was one help button in the file browser modal. Two, however, is a problem.

Is it limited to the file browser? Are any other non-modal pages affected? Any other modal affected?

**Comment by BinkaDroid** [comment]
Until now, I only noticed it in the file browser (where no button should be visible, as it is a 'footer').

I just had a look at addons which use `register('page_end')`:
![pe](https://user-images.githubusercontent.com/39274494/56060774-14e45100-5d68-11e9-948d-984aa04886ec.png)

None of theese should get executed more than once per hole page (including modals). Even if it doesn't affect the visible final layout of the page, it at least produces non-valid html if the addons insert multipe ID's with the same name ...

I don't think any addon-developer would expect a page_end-hook to get executed more than once per hole page-render, so I think friendica should't execute them, when tey are called from within a modal ...

**Comment by MrPetovan** [comment]
Ha, good point. Let's review the places the `page_end` hooks are called.

**Comment by MrPetovan** [comment]
So, preventing the `page_end` hooks to be called in asynchronously loaded content was easy. Figuring out why it used to be called twice in the file browser was easy. Fixing the latter won't be that easy, because the file browser is built with the assumption that its content will be added inside itself.

## Technical details

At first the file browser is loaded like this in the modal:
```html
<div id="jot-fbrowser-wrapper" class="" aria-labelledby="jot-browser-link" role="tabpanel" aria-hidden="false">
	<!--
		This is the template used by mod/fbrowser.php
	-->
	<style>
		#buglink_wrapper{display:none;} /* hide buglink. only in this page */
	</style>

	<div class="fbrowser image">
		<div class="fbrowser-content">
			(Images, buttons, form elements)
		</div>

		<div class="profile-rotator-wrapper" aria-hidden="true" style="display: none;">
			<i class="fa fa-circle-o-notch fa-spin" aria-hidden="true"></i>
		</div>
	</div>
</div>
```
If a folder link is clicked however, the content is loaded inside `.fbrowser` and then it looks like this:
```html
<div id="jot-fbrowser-wrapper" class="" aria-labelledby="jot-browser-link" role="tabpanel" aria-hidden="false">
	<!--
		This is the template used by mod/fbrowser.php
	-->
	<style>
		#buglink_wrapper{display:none;} /* hide buglink. only in this page */
	</style>

	<div class="fbrowser image">
		<!--
			This is the template used by mod/fbrowser.php
		-->
		<style>
			#buglink_wrapper{display:none;} /* hide buglink. only in this page */
		</style>

		<div class="fbrowser image">
			<div class="fbrowser-content">
				(Buttons, Images, Form elements...)
			</div>

			<div class="profile-rotator-wrapper" aria-hidden="true" style="display: none;">
				<i class="fa fa-circle-o-notch fa-spin" aria-hidden="true"></i>
			</div>
		</div>
	</div>
</div>
```

However, when I changed the target element for content inclusion to `#jot-fbrowser-wrapper`, the file browser navigation links stopped working after the asynchronous load.

So it isn't a straightforward change, the file browser is built on this oddity which made the `page_end` content display twice in the file browser.

Any idea @rabuzarus ?

**Comment by BinkaDroid** [comment]
Ahh, there's already a "dirty hack" in there, to prevent the bug-icon from showing up there ;-)
`#buglink_wrapper{display:none;} /* hide buglink. only in this page */`

Just "fixing" the file browser isn't the right solution I think ...
As of now, I don't see lots of modal/ asynchronous stuff besides fbrowser in friendica, but this could change in the future and can cause lots of trouble, if certain hooks are executed again after an initial page is completley rendered (eg` head`, `page_header`, `page_end`)

**Comment by MrPetovan** [comment]
Yes, there are two parts to this issue, the `page_end` hook that can easily be prevented for asynchronous calls, and the weird file browser recursive behavior. I'll submit the first part soon and I'll see about the second part later.

**Comment by nupplaphil** [comment]
Fixed with #7003
